galsim/config/extra.py

```diff
@@ -0,0 +1,412 @@
+        access self.data[k], where k = base['image_num'] - base['start_image_num'] is the
+        appropriate index to use for this image.
+
+        @param index        The index in self.data to use for this image.  This isn't the image_num
+                            (which can be accessed at base['image_num'] if needed), but rather
+                            an index that starts at 0 for the first image being worked on and
+                            goes up to nimages-1.
+        @param obj_nums     The object numbers that were used for this image.
+        @param config       The configuration field for this output object.
+        @param base         The base configuration dict.
+        @param logger       If given, a logger object to log progress. [default: None]
+        """
+        pass
+
+    def finalize(self, config, base, logger):
+        """Perform any final processing at the end of all the image processing.
+
+        This function will be called after all images have been built.
+
+        @param config       The configuration field for this output object.
+        @param base         The base configuration dict.
+        @param logger       If given, a logger object to log progress. [default: None]
+        """
+        pass
+
+    def writeFile(self, file_name, config, base, logger):
+        """Write this output object to a file.
+
+        @param file_name    The file to write to.
+        @param config       The configuration field for this output object.
+        @param base         The base configuration dict.
+        @param logger       If given, a logger object to log progress. [default: None]
+        """
+        pass
+
+    def writeHdu(self, config, base, logger):
+        """Write the data to a FITS HDU with the data for this output object.
+
+        @param config       The configuration field for this output object.
+        @param base         The base configuration dict.
+        @param logger       If given, a logger object to log progress. [default: None]
+
+        @returns an HDU with the output data.
+        """
+        raise NotImplemented("The %s class has not overridden writeHdu."%self.__class__)
+
+
+def RegisterExtraOutput(key, builder):
+    """Register an extra output field for use by the config apparatus.
+
+    The builder parameter should be a subclass of galsim.config.ExtraOutputBuilder.
+    See that class for the functions that should be defined and their signatures.
+    Not all functions need to be overridden.  If nothing needs to be done at a particular place
+    in the processing, you can leave the base class function, which doesn't do anything.
+
+    @param key              The name of the output field in config['output']
+    @param builder          A builder object to use for building the extra output object.
+                            It should be an instance of a subclass of ExtraOutputBuilder.
+    """
+    valid_extra_outputs[key] = builder
```

Comment:
I think I'll pass on this recommendation.  No offense.  :)  I'm not super familiar with decorators yet, and I'm worried that this will be intimidating to other users who want to use this registration functionality, who may be even less familiar with decorator syntax.

Also, I think as written, this will assign the _class_ to the `valid_extra_outputs` dict rather than an _instance_ of that class.  For the `ExtraOutputBuilder`s, we could do it that way: have the class be the thing that is registered and make an instance later when we need it.  

But some of the other things I register use the same class with different initialization parameters for different types.  So I decided it made more sense to have everything register instances rather than classes to be consistent.  I think in the decorator syntax, this would look like:
```
def RegisterExtraOutput(key):
    def decorate(builder):
        valid_extra_outputs[key] = builder

class BadPixBuilder(ExtraOutputBuilder):
    ...

@RegisterExtraOutput("badpix")
BadPixBuilder()
```
which doesn't seem like a big improvement over the existing syntax using the function call.